text/0585-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,793 @@
+-  factoryFor(fullName: string, options: LookupOptions): FactoryManager;
++  factoryFor(identifier: Identifier, options: LookupOptions): FactoryManager;
+
+-  hasRegistration(fullName: string): boolean;
++  hasRegistration(identifier: Identifier): boolean;
+
+-  inject(factoryNameOrType: string, property: string, injectionName: string): void;
++  inject(factory: Identifier | FactoryTypeIdentifier, property: string, injection: Identifier): void;
+
+-  lookup(fullName: string, options?: LookupOptions): any;
++  lookup(identifier: Identifier, options?: LookupOptions): any;
+
+-  register(fullName: string, factory: any, options?: RegisterOptions): void;
++  register(identifier: Identifier, factory: any, options?: RegisterOptions): any;
+
+-  registerOptions(fullName: string, options: RegisterOptions): void;
++  registerOptions(identifier: Identifier, options: RegisterOptions): any;
+
+-  registerOptionsForType(fullName: string, options: RegisterOptions): void;
++  registerOptionsForType(identifier: FactoryTypeIdentifier, options: RegisterOptions): any;
+
+-  registeredOption(fullName: string, optionName: string): RegisterOptions;
++  registeredOption(identifier: Identifier, optionName: string): RegisterOptions;
+
+-  registeredOptions(fullName: string): RegisterOptions;
++  registeredOptions(identifier: Identifier): RegisterOptions;
+
+-  registeredOptionsForType(type: string): RegisterOptions;
++  registeredOptionsForType(type: FactoryTypeIdentifier): RegisterOptions;
+
+-  resolveRegistration(type: string): Factory;
++  resolveRegistration(identifier: Identifier): Factory;
+
+-  unregister(fullName: string): void;
++  unregister(identifier: Identifier): void;
+ }
+```
+
+### Service and Controller Injections
+
+The other major examples of registry interactions are service and controller injections. This RFC proposes to normalize both to support the same identifier-based lookup as the `Owner` APIs.
+
+We introduce an `InjectionIdentifier` type, representing the service or controller to be injected:
+
+```ts
+interface InjectionIdentifier {
+  name: string;
+  namespace?: string;
+}
+```
+
+The `inject` (whether `as service` or `as controller`) function now accepts *either* a string representing the  *or* an `InjectionIdentifier`:
+
+- **Service:**
+    ```ts
+    export function inject(name: string): Service;
+    export function inject(identifier: InjectionIdentifier): Service;
+    ```
+
+- **Controller:**
+    ```ts
+    export function inject(name: string): Controller;
+    export function inject(identifier: InjectionIdentifier): Controller;
+    ```
+
+The `'<namespace>@<name>'` form is deprecated and will be removed at 4.0. The no-argument form, and the name-only string form are still allowed. Users *may* invoke using `{ name: '<name>' }` but are not required to.
+
+- **Good** (using services as an example):
+
+    ```js
+    import { inject as service } from '@ember/service';
+
+    export default class Example {
+      @inject foo;
```

Comment:
import renamed, but used without the rename

Reply:
Ah, nice catch on these; thank you! (No matter how many times you proof-read, as *soon* as it goes up all the mistakes are obvious. 😂)